### DISCUSSION TITLE
Could we get rid of the code implementing the optimization to share finally block ?

### DISCUSSION THREAD

**Discussion by srikanth-sankaran** [eclipse-jdt/eclipse.jdt.core#3246] [General]
Background: https://bugs.eclipse.org/bugs/show_bug.cgi?id=404146

More than a decade ago, we disabled the optimization to "share"  code emitted for finally blocks. This is because, it emerged that our infrastructure has some gaps in analysis that would lead to bad code generation in some cases. The fundamental problem was that, the precise knowledge of the initialization status of secret variables synthesized by various parts of the compiler was not known to the flow analysis phase. (Note: The number of secret variables have increased in the past decades with the implementation of many JEPs in the interim resulting in these to increase in number)

So we decided to disable this optimization but per suggestion in https://bugs.eclipse.org/bugs/show_bug.cgi?id=404146#c21 we still had this behavior under an option `org.eclipse.jdt.internal.compiler.impl.CompilerOptions.OPTION_ShareCommonFinallyBlocks`

I believe this option was never documented and is not exposed in the UI (@noopur2507  - can you confirm ? TIA) 

The only tests that seem to care are 
1. `org.eclipse.jdt.core.tests.compiler.regression.XLargeTest.test015()` 

which will fail with an error `The code of method foo(int) is exceeding the 65535 bytes limit` - if this sharing behavior is completely removed - this is as predicted by https://bugs.eclipse.org/bugs/show_bug.cgi?id=404146#c15.

As the same comment mentions, javac doesn't implement this optimization and so **compiling this test case with javac also fails** with:
`X.java:3: error: code too large for try statement`

2. `org.eclipse.jdt.core.tests.compiler.regression.TryStatementTest.test042()` : This failure is _not a real functionality test_. It is checking that the bytecodes produced do share the finally blocks

I would like to hear any objections to getting rid of this code against these considerations:

    - Behavior is disabled for more than a decade, nobody has complained about it
    - Can be enabled only by undocumented option, not exposed in the UI
    - We have never had to recommend this option to anyone: A bugzilla and Github search for OPTION_ShareCommonFinallyBlocks  brings up nothing.
    - Downside is similar to javac's - they also have the bloat issue in the worst case.

The code involved is not huge, but it is like a death by a thousand cuts if we don't clear up the clutter in various places each being small.

@stephan-herrmann @jarthana @mpalat @iloveeclipse : Thoughts ?

**Comment by noopur2507** [comment]
> I believe this option was never documented and is not exposed in the UI (@noopur2507 - can you confirm ? TIA)

I do not see any corresponding UI/code for this option.

**Comment by iloveeclipse** [comment]
No objections from me.

**Reply by stephan-herrmann** [reply]
Same here.

**Comment by srikanth-sankaran** [comment]
Thanks!  I will merge a patch to master on Friday unless in the interim I hear any counter opinions.

**Comment by srikanth-sankaran** [comment]
PR here: https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3281
